### PR TITLE
feat(googleOauth): add readCommentsOnDoc action

### PR DESCRIPTION
**Pull request by MattBetancourt**
## Description

This PR introduces the `readCommentsOnDoc` action for the `google-oauth` provider, enabling agents to fetch comments (with exact highlighted anchor text) from both native Google Docs and raw DOCX files stored in Drive.

## Architecture

```mermaid
flowchart TD
  A["readCommentsOnDoc(documentId)"] --> B["Fetch Drive metadata<br/>mimeType, size"]
  B --> S1["Safety checks<br/>auth required, supported type only,<br/>50MB file/export ceiling"]
  S1 --> C{"File type?"}

  C -->|Uploaded DOCX| D["Download DOCX from Drive"]
  D --> S2["DOCX parser guardrails<br/>20MB per XML part decompression cap,<br/>XML entities disabled,<br/>max XML nesting depth"]
  S2 --> E["Parse OpenXML<br/>comments.xml, document.xml,<br/>headers, footers"]
  E --> F["Extract comments, replies,<br/>exact anchored text, inline images,<br/>document position"]
  F --> Z["Filter + sort<br/>resolved handling, document order"]
  Z --> OUT["Return comments"]

  C -->|Native Google Doc| G["List Drive comments<br/>authoritative metadata"]
  G --> S3["Pagination guardrail<br/>encoded page tokens,<br/>100 page cap, warning on truncation"]
  S3 --> H["Export as DOCX<br/>supportsAllDrives=true"]
  H --> S2
  E --> I["Deterministic join<br/>author + content + timestamp seconds"]
  I --> J["Attach exact anchors<br/>anchorConfidence: exact/none"]
  J --> Z

  H -->|Export fails| K["Fallback to Drive comments only"]
  K --> L["Return warning<br/>anchorConfidence: none"]
  L --> OUT

  C -->|Unsupported type| X["Return error"]
```

* **Dual-path Strategy**: The action attempts to fetch Google Docs via their native Drive comments API (PATH A). However, the Drive API does not return the highlighted text that a comment is anchored to. Therefore, the action silently (and transparently to the caller), exports the document to DOCX, and parses the OpenXML to locate the exact anchor text.
  * **Raw DOCX Files**: Downloads the file directly and parses the OpenXML to extract comments, anchor text, and reply threading.
* **Deterministic Joining**: Uses a highly efficient O(n+m) Map-based index (matching on Author, Content, and Truncated Date) to join the Drive API comment metadata with the exact document anchor positions extracted from the DOCX export.
* **Rich Edge-Case Handling**: The DOCX OpenXML parser safely handles hyperlinks, inline images (extracting alt text), smart chips, mixed text styles, nested comment ranges, lists, tables, and headers/footers.
* **Pagination Safety**: Added safe Drive API `comments.list` pagination with encoded page tokens and a 100-page cap to prevent unbounded API consumption on massive documents.

## Architectural Decisions & Trade-offs

* **New Dependencies (`jszip`, `fast-xml-parser`)**: Added these two packages to handle the in-memory parsing of the downloaded/exported DOCX binaries. `fast-xml-parser` is used over the built-in `DOMParser` for speed and precise `preserveOrder` capabilities (required to track reading-order positions in OpenXML).
  * **Safety Measures Added**: Because we are introducing these new parsing dependencies, we implemented strict security guardrails. This includes a 20MB decompression limit to prevent Zip Bombs, a 50MB arraybuffer ceiling to protect Node's heap, and disabling XML entity expansion to mitigate Billion Laughs DoS attacks.
* **Undocumented Google Behavior**: Google does not officially document that `files.export` (DOCX) preserves comment anchors. This implementation relies on reverse-engineering OpenXML (`w:commentRangeStart` / `End`). Because this is not a formal contract, extensive synthetic integration tests are included to detect if Google ever changes this behavior.
* **Deterministic Join Strategy**: Drive API comment IDs do not match the local integer IDs generated in the DOCX export. To merge the two, we use a deterministic join requiring an exact match on:
  1. Author display name
  2. Comment body text
  3. Truncated timestamp (Drive API uses milliseconds; DOCX export drops them, so we join on truncated seconds).

## Testing

* Comprehensive Jest unit tests added in `tests/google-oauth/testReadCommentsOnDoc.test.ts` and `tests/google-oauth/testReadCommentsOnDocAction.test.ts`.
* Included synthetic DOCX generation tests that validate complex parsing edge cases (e.g., duplicated text anchors, multi-paragraph highlights, inline images, reply threading).
* Validated the deterministic join algorithm between Drive API milliseconds and DOCX truncated seconds, plus Drive pagination edge cases.
* Passed all linter (`eslint`) and TypeScript build checks.

## Documentation

No external documentation updates are needed for this change. The new action is fully defined in `schema.yaml` and will be automatically included in the auto-generated API references. The internal DOCX fallback behavior is transparent to callers and does not require explicit documentation.

## Changes Made

* Added: `src/actions/providers/google-oauth/readCommentsOnDoc.ts`
* Updated: `src/utils/google.ts` (added DOCX parsing utilities and join logic)
* Updated: `src/actions/schema.yaml` (defined `readCommentsOnDoc` with `drive.readonly` scope)
* Updated: `src/actions/actionMapper.ts` (registered the action)
* Generated: Updated autogen types and templates via `npm run generate:types`

### CODE REVIEW

File: src/actions/providers/google-oauth/readCommentsOnDoc.ts [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+            author: r.author ? { displayName: r.author.displayName, emailAddress: r.author.emailAddress } : undefined,
+          })),
+        };
+      });
+
+      // Try to get precise anchors from exported DOCX
+      try {
+        const exportRes = await axiosClient.get(
+          `${GDRIVE_BASE_URL}${encodeURIComponent(documentId)}/export?mimeType=application/vnd.openxmlformats-officedocument.wordprocessingml.document`,
+          {
```
[the remark this answers is not in the document]
Comment (MattBetancourt) [reply src/actions/providers/google-oauth/readCommentsOnDoc.ts:201]: Good catch — fixed in the latest commit. We verified via live API testing that `supportsAllDrives=true` is accepted as a valid parameter on `files.export` (returns `HTTP 200`), and the DOCX export URL now includes it. Also restored the parameter to the export fallback paths in `getGoogleDocContent`, `getGoogleSheetContent`, and `getGoogleSlidesContent` which had it inadvertently dropped.

**Comment by MattBetancourt** [comment]
Addressing [the zip-bomb guard bypass finding](https://github.com/Credal-ai/actions-sdk/pull/571#discussion_r3321320194) — fixed in the latest commit. The `|| 0` fallback was replaced with an explicit `undefined` check: if `_data?.uncompressedSize` is missing (encrypted entry, future JSZip version, or adversarially zeroed local file header), the entry is now rejected rather than silently let through. Legitimate zero-byte files are still allowed since `0 > 20MB` is false.
